### PR TITLE
Removed type casts

### PR DESCRIPTION
**Pull request by AndrewMusgrave**
### WHY are these changes introduced?

Let us be even more type-safe

### WHAT is this pull request doing?

Remove the keyword `as` from the majority of production files by using guards.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}

```

</details>

### 🎩 checklist

* [ ] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated the component's `README.md` with documentation changes
* [ ] [Tophatted documentation](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting%20documentation.md) changes in the style guide
* [ ] For visual design changes, pinged one of @ HYPD, @ mirualves, @ sarahill, or @ ry5n to update the Polaris UI kit

### CODE REVIEW

File: src/components/TrapFocus/tests/TrapFocus.test.tsx [hunk 1]
```diff
@@ -180,6 +180,7 @@ describe('<TrapFocus />', () => {
       const event: Partial<FocusEvent> = {
         relatedTarget: externalDomNode,
         srcElement: externalDomNode,
+        target: document.createElement('div'),
```
Comment (AndrewMusgrave) [inline src/components/TrapFocus/tests/TrapFocus.test.tsx]: This test was relying on a target

File: src/components/Image/Image.tsx [hunk 2]
```diff
@@ -10,6 +10,7 @@ export type CrossOrigin = 'anonymous' | 'use-credentials' | '' | undefined;
 export interface ImageProps extends React.HTMLProps<HTMLImageElement> {
   alt: string;
   source: string;
+  crossOrigin?: CrossOrigin;
```
Comment (AndrewMusgrave) [inline src/components/Image/Image.tsx]: Rather than casting, let us add it to our interface!

File: src/components/Modal/Modal.tsx [hunk 3]
```diff
@@ -159,15 +159,7 @@ class Modal extends React.Component<CombinedProps, State> {
...
     ) {
-      if (isIframeModal(transformedProps)) {
-        (this.appBridgeModal as AppBridgeModal.ModalIframe).set(
-          transformedProps,
-        );
-      } else {
-        (this.appBridgeModal as AppBridgeModal.ModalMessage).set(
-          transformedProps,
-        );
-      }
```
Comment (AndrewMusgrave) [inline src/components/Modal/Modal.tsx]: Couldn't find a reason for this `if/else`. Seems to be executing the same code.

File: src/components/Checkbox/Checkbox.tsx [hunk 4]
```diff
@@ -127,7 +127,7 @@ export const Checkbox = React.forwardRef<CheckboxHandles, CheckboxProps>(
     const isChecked = !isIndeterminate && Boolean(checked);
 
     const indeterminateAttributes = isIndeterminate
-      ? {indeterminate: 'true', 'aria-checked': 'mixed' as 'mixed'}
+      ? {indeterminate: 'true', 'aria-checked': 'mixed' as const}
```
Comment (dleroux) [inline src/components/Checkbox/Checkbox.tsx]: curious whats the benefit here?

File: src/components/ColorPicker/components/Slidable/Slidable.tsx [hunk 5]
```diff
@@ -185,3 +179,13 @@ export class Slidable extends React.PureComponent<SlidableProps, State> {
     onChange({x: offsetX, y: offsetY});
   };
 }
+
+function isMouseMoveEvent(event: MouseEvent | TouchEvent): event is MouseEvent {
```
Comment (dleroux) [inline src/components/ColorPicker/components/Slidable/Slidable.tsx]: 😍

File: src/components/Checkbox/Checkbox.tsx (hunk 4 above)
Comment (AndrewMusgrave) [reply src/components/Checkbox/Checkbox.tsx]: [ts docs](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html#const-assertions) explain it well 😄

File: src/components/Pagination/Pagination.tsx [hunk 6]
```diff
@@ -144,6 +144,7 @@ export function Pagination({
       nextButton
     );
 
+  const previousHandler = onPrevious || noop;
```
Comment (BPScott) [inline src/components/Pagination/Pagination.tsx]: What do you think to setting noop as the default value when we destructure the arguments of the Pagination function?

```
onNext = noop,
onPrevious = noop,
```

Then the handleCallback call would continue to be `handleCallback(onPrevious)`.

We do that in a bunch of other places, e.g. ChoiceList onChange (search for `= noop`)

File: src/components/ColorPicker/components/Slidable/Slidable.tsx [hunk 7]
```diff
@@ -185,3 +184,13 @@ export class Slidable extends React.PureComponent<SlidableProps, State> {
...
   };
 }
+
+function isMouseEvent(
+  event: Event | MouseEvent | React.MouseEvent | React.TouchEvent,
+): event is MouseEvent {
+  return event.type === 'mousemove' || event.type === 'mousedown';
+}
+
+function isTouchEvent(event: Event | TouchEvent): event is TouchEvent {
```
[the remark this answers is not in the document]
Comment (AndrewMusgrave) [reply src/components/ColorPicker/components/Slidable/Slidable.tsx]: Already removed the function 😄

File: src/components/ColorPicker/components/Slidable/Slidable.tsx [hunk 8]
```diff
@@ -185,3 +184,13 @@ export class Slidable extends React.PureComponent<SlidableProps, State> {
     onChange({x: offsetX, y: offsetY});
   };
 }
+
+function isMouseEvent(
+  event: Event | MouseEvent | React.MouseEvent | React.TouchEvent,
```
[the remark this answers is not in the document]
Comment (AndrewMusgrave) [reply src/components/ColorPicker/components/Slidable/Slidable.tsx]: Nice catch! Event should work 👍

File: src/components/Pagination/Pagination.tsx (hunk 6 above)
Comment (AndrewMusgrave) [reply src/components/Pagination/Pagination.tsx]: Unfortunately, that'll break some conditional logic we have for deciding whether to or not to render a listener `(previousURL || onPrevious) &&`

**Pull request merged**